k3s-io/k3s
# Add certificate expiry check, events, and metrics

#### Proposed Changes ####

* Moves client and server cert listing code into utils/service (open to suggestions on a better place/name for this)
* Fix `k3s certificate rotate` failing on agents due to missing server token file
* Adds `k3s certificate check` command, to check and print the status of server and agent certificates
* Adds periodic automatic checking of certificates; creates warning events attached to the node for certs that are expired or about to expire, and updates metrics

#### Types of Changes ####

enhancement

#### Verification ####

* Start K3s with `CATTLE_NEW_SIGNED_CERT_EXPIRATION_DAYS=30` in service environment to trigger short-lived certs
* Run new `k3s certificate check` command
* Check for events


#### Testing ####


#### Linked Issues ####

* https://github.com/k3s-io/k3s/issues/9785
* https://github.com/k3s-io/k3s/issues/9742

#### User-Facing Change ####

```release-note

```

#### Further Comments ####
